README.md
<p align="center">
  <a href="https://symbology.dev/">
    <img src="https://symbology.dev/assets/hero.svg" width="159" height="95" />
  </a>

  <h1 style="text-align: center;" align="center">Symbology</h1>
</p>

<p align="center">A Node.js module that generates barcode images. Supports 50+ different 1D or 2D symbologies in png, eps, or svg formats.</p>

<p align="center">
  <a href="https://app.codecov.io/gh/jshor/symbology"><img
    src="https://img.shields.io/codecov/c/gh/jshor/symbology?style=for-the-badge"
    alt="Code coverage"
  /></a> <a href="https://github.com/jshor/symbology/actions/workflows/merge.yml"><img
    src="https://img.shields.io/github/actions/workflow/status/jshor/symbology/merge.yml?branch=main&style=for-the-badge"
    alt="Build status"
  /></a> <a href="https://npmjs.com/package/symbology"><img
    src="http://img.shields.io/npm/v/symbology.svg?style=for-the-badge"
    alt="npm version"
  /></a>
</p>

## Introduction

This Node.js module will allow you to generate over 50+ different types of 1D or 2D symbologies, including barcodes for books, grocery, shipping carriers, healthcare, and international codes.

It can create a PNG, SVG, or EPS image file, or return a string containing SVG, PostScript, or base64-encoded PNG data.

## Documentation

[Read the docs →](https://symbology.dev/docs/)

## Quick start

```sh
yarn add symbology
```

## Quick Examples

### Code 11 Example

```ts
import { SymbologyType, createStream } from 'symbology'

(async () => {
  const { data } = await createStream({
    symbology: SymbologyType.CODE11
  }, '8765432164')

  console.log('Result: ', data)
})()
```

This will log:

```json
{
  "data": "data:image/png+data;base64,PHN [...] eFd==",
  "message": "Symbology successfully created.",
  "code": 0
}
```

And the base64 PNG generated will look like:

![code 11](https://symbology.dev/assets/barcodes/barcode_14.png)

### MaxiCode Example

```ts
import { SymbologyType, createFile } from 'symbology'

(async () => {
  const { data } = await createFile({
    symbology: SymbologyType.MAXICODE,
    option1: 2,
    primary: '999999999840012',
    fileName: 'maxiCodeExample.svg'
  }, 'Secondary Message Here')

  console.log('Result: ', data)
})()
```

This creates `maxiCodeExample.svg` which looks like:

![MaxiCode](https://symbology.dev/assets/barcodes/barcode_47.png)

### USPS Example

```ts
import { SymbologyType, createFile } from 'symbology'

(async () => {
  const { data } = await createFile({
    symbology: SymbologyType.ONECODE
    fileName: 'uspsExample.eps'
  }, '01234567094987654321-01234')

  console.log('Result: ', data)
})()
```

This creates `uspsExample.eps` which looks like:

![USPS](https://symbology.dev/assets/barcodes/barcode_42.png)

## License

[GPL-3](LICENSE.md).

## Windows

```sh
npm config set msvs_version 2022 --global
```


CHANGELOG.md
## [4.0.1](https://github.com/jshor/symbology/compare/v4.0.0...v4.0.1) (2023-11-22)



# [4.0.0](https://github.com/jshor/symbology/compare/v3.0.13...v4.0.0) (2023-10-15)


### Features

* **node:** upgrades to support Node 18, 20 [#120](https://github.com/jshor/symbology/issues/120) ([3b7904d](https://github.com/jshor/symbology/commit/3b7904dddf8f1e2abc50b8910423d8fba4770e7e))


### BREAKING CHANGES

* **node:** This drops support for Node.js v14.



## [3.0.13](https://github.com/jshor/symbology/compare/v3.0.12...v3.0.13) (2022-11-28)



## [3.0.12](https://github.com/jshor/symbology/compare/v3.0.11...v3.0.12) (2022-11-28)



## [3.0.11](https://github.com/jshor/symbology/compare/v3.0.10...v3.0.11) (2022-11-27)



## [3.0.10](https://github.com/jshor/symbology/compare/v3.0.9...v3.0.10) (2022-11-27)


### Bug Fixes

* **bin:** updates binary to install as-needed ([86ba08c](https://github.com/jshor/symbology/commit/86ba08c29dbd0c3933637dc4f825ec4cd6a737f5))



## [3.0.9](https://github.com/jshor/symbology/compare/v3.0.8...v3.0.9) (2022-11-27)


### Bug Fixes

* **svg:** adds fix for SVG/EPS crashes [#96](https://github.com/jshor/symbology/issues/96) ([300a007](https://github.com/jshor/symbology/commit/300a007e0073f32b933a1ced89653dbb40a10ed2))



## [3.0.8](https://github.com/jshor/symbology/compare/v3.0.7...v3.0.8) (2021-11-20)



## [3.0.7](https://github.com/jshor/symbology/compare/v3.0.6...v3.0.7) (2021-11-20)


### Bug Fixes

* **deps:** corrects nan dependency ([45b17cd](https://github.com/jshor/symbology/commit/45b17cdedc0b385b5e939bb67ceba0894037884b))



## [3.0.6](https://github.com/jshor/symbology/compare/v3.0.5...v3.0.6) (2021-09-04)



## [3.0.5](https://github.com/jshor/symbology/compare/v3.0.4...v3.0.5) (2021-08-28)



## [3.0.4](https://github.com/jshor/symbology/compare/v3.0.3...v3.0.4) (2021-06-29)



## [3.0.3](https://github.com/jshor/symbology/compare/v3.0.2...v3.0.3) (2021-06-22)



## [3.0.2](https://github.com/jshor/symbology/compare/v3.0.1...v3.0.2) (2021-06-15)


### Bug Fixes

* corrects syntax error [#71](https://github.com/jshor/symbology/issues/71) ([bcf48d2](https://github.com/jshor/symbology/commit/bcf48d25230a4eacd72ed3356adfdc7aa6e9273a))



## [3.0.1](https://github.com/jshor/symbology/compare/v3.0.0...v3.0.1) (2021-06-05)



# [3.0.0](https://github.com/jshor/symbology/compare/v2.1.6...v3.0.0) (2021-06-01)


### Features

* v3 ([6571397](https://github.com/jshor/symbology/commit/6571397b529e25207da8aa131083beed17c02b20))


### BREAKING CHANGES

* drops support for node v12 and v13.



## [2.1.6](https://github.com/jshor/symbology/compare/v2.1.5...v2.1.6) (2021-04-30)



## [2.1.5](https://github.com/jshor/symbology/compare/v2.1.4...v2.1.5) (2021-04-10)



## [2.1.4](https://github.com/jshor/symbology/compare/v2.1.2...v2.1.4) (2021-04-08)


### Bug Fixes

* **build:** fixes compatibility of module exports and npm ([45ebe3b](https://github.com/jshor/symbology/commit/45ebe3b57fd7e187d5ba624525063938f5cc490a))



## [2.1.2](https://github.com/jshor/symbology/compare/v2.1.1...v2.1.2) (2021-04-04)



## [2.1.1](https://github.com/jshor/symbology/compare/v2.1.0...v2.1.1) (2021-04-01)



# [2.1.0](https://github.com/jshor/symbology/compare/v2.0.2...v2.1.0) (2021-04-01)


### Features

* **ts:** updates package to use TS ([2c4740a](https://github.com/jshor/symbology/commit/2c4740a2dbc47ab79864a451cfaf8d049129d128))



## [2.0.2](https://github.com/jshor/symbology/compare/v2.0.1...v2.0.2) (2021-03-31)


### Bug Fixes

* **lib:** fixes EOF and npmignore issues [#55](https://github.com/jshor/symbology/issues/55) ([085d33b](https://github.com/jshor/symbology/commit/085d33bfdf0540cb5138c3b3bf6f983ae0ba5a16))



## [2.0.1](https://github.com/jshor/symbology/compare/v2.0.0...v2.0.1) (2021-03-29)



# [2.0.0](https://github.com/jshor/symbology/compare/v1.2.3...v2.0.0) (2021-03-29)


### Bug Fixes

* adds patch system to apply stdout fixes ([d0ac5bc](https://github.com/jshor/symbology/commit/d0ac5bcab4be9d18dcb32ca71f9e30174d44f46d))
* adds support for Node v12+ ([bffc800](https://github.com/jshor/symbology/commit/bffc800e0bd57e4cf48913bf0d79c33517ab1fd3))
* **appveyor:** update script to include debug msg and wrap conditionals ([#26](https://github.com/jshor/symbology/issues/26)) ([2859265](https://github.com/jshor/symbology/commit/285926561bc0d9b7e62e26b7c1d188b0fef0567e))
* **binding:** updates lib to properly convert bitmap chars ([51c0990](https://github.com/jshor/symbology/commit/51c0990b08bcb5aff2b38dd45113fc5a6e1ae365))
* powershell bug ([#27](https://github.com/jshor/symbology/issues/27)) ([c8e6fdf](https://github.com/jshor/symbology/commit/c8e6fdf3f70a04384dc97f471e042220dddc7afc))
* **unix:** removes remaining png.c refs; enables pipe() for non-Win32 ([7131d1c](https://github.com/jshor/symbology/commit/7131d1c173a2dff7f86a8d7a07cb4dee8b4dee0b))
* **version:** adds missing C file ref, version mgmt ([#37](https://github.com/jshor/symbology/issues/37)) ([0903303](https://github.com/jshor/symbology/commit/090330333cdefdedd9e72eccf0e94c4524daf6ec))


### Features

* **alpha:** implements alpha channel support [#35](https://github.com/jshor/symbology/issues/35) ([993699b](https://github.com/jshor/symbology/commit/993699b9468d0d00ac413fdba8e1f02da1f7318a))
* **encoding:** adds encoding support for ECI and input modes ([#30](https://github.com/jshor/symbology/issues/30)) ([e9ff0d0](https://github.com/jshor/symbology/commit/e9ff0d086f5abd5c2c1cfa85c463909b870d2566))
* implements ability to alter dot size [#32](https://github.com/jshor/symbology/issues/32) ([32b9b42](https://github.com/jshor/symbology/commit/32b9b42775b7943e54d6586d4aaf1ec8895c29d5))
* **rotation:** implements rotations [#31](https://github.com/jshor/symbology/issues/31) ([11b69e6](https://github.com/jshor/symbology/commit/11b69e6844a85c8bb6170c0476ccb8633d5a0e9f))

enums.js
/* Tbarcode = 7 codes */
var BARCODE_CODE11 = 1,
  BARCODE_C25MATRIX = 2,
  BARCODE_C25INTER = 3,
  BARCODE_C25IATA = 4,
  BARCODE_C25LOGIC = 6,
  BARCODE_C25IND = 7,
  BARCODE_CODE39 = 8,
  BARCODE_EXCODE39 = 9,
  BARCODE_EANX = 13,
  BARCODE_EAN128 = 16,
  BARCODE_CODABAR = 18,
  BARCODE_CODE128 = 20,
  BARCODE_DPLEIT = 21,
  BARCODE_DPIDENT = 22,
  BARCODE_CODE16K = 23,
  BARCODE_CODE49 = 24,
  BARCODE_CODE93 = 25,
  BARCODE_FLAT = 28,
  BARCODE_RSS14 = 29,
  BARCODE_RSS_LTD = 30,
  BARCODE_RSS_EXP = 31,
  BARCODE_TELEPEN = 32,
  BARCODE_UPCA = 34,
  BARCODE_UPCE = 37,
  BARCODE_POSTNET = 40,
  BARCODE_MSI_PLESSEY = 47,
  BARCODE_FIM = 49,
  BARCODE_LOGMARS = 50,
  BARCODE_PHARMA = 51,
  BARCODE_PZN = 52,
  BARCODE_PHARMA_TWO = 53,
  BARCODE_PDF417 = 55,
  BARCODE_PDF417TRUNC = 56,
  BARCODE_MAXICODE = 57,
  BARCODE_QRCODE = 58,
  BARCODE_CODE128B = 60,
  BARCODE_AUSPOST = 63,
  BARCODE_AUSREPLY = 66,
  BARCODE_AUSROUTE = 67,
  BARCODE_AUSREDIRECT = 68,
  BARCODE_ISBNX = 69,
  BARCODE_RM4SCC = 70,
  BARCODE_DATAMATRIX = 71,
  BARCODE_EAN14 = 72,
  BARCODE_CODABLOCKF = 74,
  BARCODE_NVE18 = 75,
  BARCODE_JAPANPOST = 76,
  BARCODE_KOREAPOST = 77,
  BARCODE_RSS14STACK = 79,
  BARCODE_RSS14STACK_OMNI = 80,
  BARCODE_RSS_EXPSTACK = 81,
  BARCODE_PLANET = 82,
  BARCODE_MICROPDF417 = 84,
  BARCODE_ONECODE = 85,
  BARCODE_PLESSEY = 86,

  /* Tbarcode = 8 codes */
  BARCODE_TELEPEN_NUM = 87,
  BARCODE_ITF14 = 89,
  BARCODE_KIX = 90,
  BARCODE_AZTEC = 92,
  BARCODE_DAFT = 93,
  BARCODE_MICROQR = 97,

  /* Tbarcode = 9 codes */
  BARCODE_HIBC_128 = 98,
  BARCODE_HIBC_39 = 99,
  BARCODE_HIBC_DM = 102,
  BARCODE_HIBC_QR = 104,
  BARCODE_HIBC_PDF = 106,
  BARCODE_HIBC_MICPDF = 108,
  BARCODE_HIBC_BLOCKF = 110,
  BARCODE_HIBC_AZTEC = 112,

  /* Zint specific */
  BARCODE_AZRUNE = 128,
  BARCODE_CODE32 = 129,
  BARCODE_EANX_CC = 130,
  BARCODE_EAN128_CC = 131,
  BARCODE_RSS14_CC = 132,
  BARCODE_RSS_LTD_CC = 133,
  BARCODE_RSS_EXP_CC = 134,
  BARCODE_UPCA_CC = 135,
  BARCODE_UPCE_CC = 136,
  BARCODE_RSS14STACK_CC = 137,
  BARCODE_RSS14_OMNI_CC = 138,
  BARCODE_RSS_EXPSTACK_CC = 139,
  BARCODE_CHANNEL = 140,
  BARCODE_CODEONE = 141,
  BARCODE_GRIDMATRIX = 142,

  BARCODE_NO_ASCII = 1,
  BARCODE_BIND = 2,
  BARCODE_BOX = 4,
  BARCODE_STDOUT = 8,
  READER_INIT = 16,
  SMALL_TEXT = 32,

  /* data input modes */
  DATA_MODE = 0,
  UNICODE_MODE = 1,
  GS1_MODE = 2,
  KANJI_MODE = 3,
  SJIS_MODE = 4,

  DM_SQUARE = 100,

  /* error codes */
  ZWARN_INVALID_OPTION = 2,
  ZERROR_TOO_LONG = 5,
  ZERROR_INVALID_DATA = 6,
  ZERROR_INVALID_CHECK = 7,
  ZERROR_INVALID_OPTION = 8,
  ZERROR_ENCODING_PROBLEM = 9,
  ZERROR_FILE_ACCESS = 10,
  ZERROR_MEMORY = 11;